src/sinks/datadog/traces/stats.rs

```diff
@@ -442,33 +524,61 @@ fn extract_weight_from_root_span(spans: &[&BTreeMap<String, Value>]) -> f64 {
                 _ => 1.0,
             })
             .unwrap_or(1.0);
+
+        // found root
         if parent_id == 0 {
-            return sample_rate;
+            return weight;
         }
+
         span_ids.push(span_id);
-        parent_id_to_child_weight.insert(parent_id, sample_rate);
+
+        parent_id_to_child_weight.insert(parent_id, weight);
     }
-    // We remove all span that do have a parent
+
+    // Remove all spans that have a parent
     span_ids.iter().for_each(|id| {
         parent_id_to_child_weight.remove(id);
     });
-    // There should be only one value remaining, the weigth from the root span
+
+    // There should be only one value remaining- the weight from the root span
     if parent_id_to_child_weight.len() != 1 {
-        debug!("Didn't reliably find the root span.");
+        let trace_id = trace_id.unwrap_or_else(|| {
+            error!("Field `trace_id` should exist in the span.");
+            -1
+        });
+        emit!(DatadogTracesStatsError {
+            error_message: "Didn't reliably find the root span for weight calculation.",
+            trace_id,
+        });
     }
 
     *parent_id_to_child_weight
         .values()
         .next()
         .unwrap_or_else(|| {
-            debug!("Root span was not found.");
+            let trace_id = trace_id.unwrap_or_else(|| {
+                error!("Field `trace_id` should exist in the span.");
```

Comment:
Should this and the above `error!` be a part of the internal event?

Reply:
I think it works either way.
The reason I kept it out of the internal event is because the only usage of `trace_id` in this function, is in these error paths. If we can't find the root span, it is the main issue and the fact that the trace_id is missing from all spans is quite possibly related but I didn't think it warranted a separate internal event, since it is all part of getting the weight.

However- I think this could be improved now that I look at it again. We could have the trace_id be an `Option` in the internal event, and that way when we fix the i64 / u64 bug , this will still work.